colossalai/shardformer/layer/attn.py

```diff
@@ -471,19 +471,42 @@ def get_double_ring_groups(sp_group, inner_ring_size=None):
         inner_ring_group = None
         inter_ring_group = None
 
-        # Create inner ring groups
-        for i in range(inner_ring_size):
-            ranks = list(range(i * inner_ring_size, (i + 1) * inner_ring_size))
-            group = dist.new_group(ranks)
-            if sp_rank in ranks:
-                inner_ring_group = group
-
-        # Create inter ring groups
-        for i in range(num_rings):
-            ranks = list(range(i, sp_size, num_rings))
-            group = dist.new_group(ranks)
-            if sp_rank in ranks:
-                inter_ring_group = group
+        world_size = dist.get_world_size()
+        rank = dist.get_rank()
+        groups = int(world_size / sp_size)
+
+        if tp_size > 1:
+            for group_id in range(groups):
```

Comment:
Not totally sure, but it should look like this?
```
world_size = 16
sp_size = 4
outer_ring_size = 2
inner_ring_size = sp_size // outer_ring_size
tp_size = 2
total_inner_rings = world_size // (outer_ring_size * tp_size) # loop through groups of size (inner_ring * tp) 
total_inner_size = tp_size * inner_ring_size
for j in range(total_inner_rings):
    # inside each group, duplicate tp group inner_sp_size times
    for k in range(tp_size):
        print(f"inner ring ranks: {list(range(k + j * total_inner_size, (j + 1) * total_inner_size, tp_size))}")
print("---------------------------------")


sp_tp_size = total_inner_size * outer_ring_size
n_groups = world_size // sp_tp_size # dp * pp
print(f"n_groups: {n_groups}")
for i in range(0, n_groups):
    start = i * sp_tp_size
    end = (i + 1) * sp_tp_size
    for j in range(outer_ring_size):    
        for k in range(tp_size):
            print(f"inter ring ranks: {list(range(start + k + j * tp_size, end, total_inner_size))}")
```
<img width="313" alt="image" src="https://github.com/user-attachments/assets/6eeacb11-3494-40e0-976c-7e22989fa31a">
<img width="579" alt="image" src="https://github.com/user-attachments/assets/54aa596f-4f60-4f60-8ffc-e0a5f3025c4b">

Reply:
Yes, I use 4 gpus in two nodes separately in the example ,see as https://arxiv.org/pdf/2406.18485. 
However, I believe there is an issue with the current algorithm, as such a rank group cannot proceed with the next step of communication. Additionally, in the CI tests, the author only tested it with 4 GPUs, so this PR may need to be closed for now.

Reply:
If use this config, which is :
inner group:[0, 1, ][2, 3],[4, 5],[ 6, 7],
inter group: [0,4][1,5][2,6][3,7]，
the communication can be proceed and can be training, seems like the current algorithm need the inner group number equals to the inter group number.
However it's wrong, the paper does not impose such requirements, the inner group can be set to 4, and the inter group to 2.

Reply:
I'm not sure I understand--the current group creation is wrong in that it doesn't consider TP and doesn't duplicate groups over dp * pp?
I fixed the code a bit, please test/modify on this and hope it works.
```
world_size = 16
sp_size = 8
outer_ring_size = 2
inner_ring_size = sp_size // outer_ring_size
tp_size = 1
total_inner_size = tp_size * inner_ring_size
total_inner_rings = world_size // total_inner_size # loop through groups of size (inner_ring * tp) 

for j in range(total_inner_rings):
    # inside each group, duplicate tp group inner_sp_size times
    for k in range(tp_size):
        print(f"inner ring ranks: {list(range(k + j * total_inner_size, (j + 1) * total_inner_size, tp_size))}")
print("---------------------------------")


sp_tp_size = total_inner_size * outer_ring_size
n_groups = world_size // sp_tp_size  # dp * pp
print(f"num dp * pp groups: {n_groups}")
for i in range(0, n_groups):
    start = i * sp_tp_size
    end = (i + 1) * sp_tp_size
    for j in range(inner_ring_size):    
        for k in range(tp_size):
            print(f"inter ring ranks: {list(range(start + k + j * tp_size, end, total_inner_size))}")
```
<img width="416" alt="image" src="https://github.com/user-attachments/assets/72e3be22-6b34-4ebc-a3ed-770b8e905edc">

Reply:
Yes but this group config have communication error with subsequent codes, that's the problem.

Reply:
okay, let me check one more time.

Reply:
Still NCCL error:
![image](https://github.com/user-attachments/assets/9b3b9459-48b9-40b5-97ff-8fb220c1e4fc)
![image](https://github.com/user-attachments/assets/46d0829e-6bed-4f70-880d-798cf612b830)
I agree with your rank group but unluckily this doesn't work, still figure it out.

Reply:
if sp = 8, world size = 16,
inner group：[0, 1, 2, 3] [4, 5, 6, 7]，
inter group：[0,4] [1,5] [2,6] [3,7]，

if tp = 2, sp = 4, world size = 16,
inner group: [0,2][1,3][4,6][5,7]，
inter group: [0,4][1,5][2,6][3,7]

Reply:
Thanks for checking and good to hear ranks are correct. 
There are two possibilities:
1. Hardware/infiniband issue. This fluctuates from time to time.
2. Using multiple NCCL process groups at the same time--intra and inter ring groups. But the original paper's implementation seems to do the same thing (https://github.com/InternLM/InternEvo/blob/e52f2ffc9acf818e8f2b1f97dfc69ceb2f06e154/internlm/model/ops/ring_flash_attn/zigzag_ring_flash_attn_with_sliding_window.py#L189)
This is as far as I can help, plz discuss further with @ver217 🙂

Reply:
Thank you for your debugging, don't worry, I just want to make sure the group rank is ok, We are in agreement now. NCCL error is something wrong with the group rank. When the tp size >1, there is no NCCL error, so at least I can use 'if else' to fix that.